File: 208.implement-trie-prefix-tree.py
#
# @lc app=leetcode id=208 lang=python3
#
# [208] Implement Trie (Prefix Tree)
#
# @lc code=start


from collections import defaultdict


class Trie:
    def __init__(self):
        self.children: dict[str, Trie] = defaultdict(Trie)
        self.is_word: bool = False

    def insert(self, word: str) -> None:
        node = self
        for c in word:
            node = node.children[c]
        node.is_word = True

    def remove(self, word: str) -> None:
        path = [self]
        for c in word:
            if not path[-1].children.get(c):
                return
            path.append(path[-1].children[c])

        if path[-1].is_word:
            return
        path[-1].is_word = False
        if len(path[-1].children) > 0:
            return

        remroot = word[-1]
        for node, c in zip(path[-2::-1], word[-2::-1]):
            if len(node.children) > 1 or node.is_word:
                del node.children[remroot]
                return
            remroot = c

    def search(self, word: str) -> bool:
        node = self
        for c in word:
            if not node.children.get(c):
                return False
            node = node.children[c]
        return node.is_word

    def startsWith(self, prefix: str) -> bool:
        node = self
        for c in prefix:
            if not node.children.get(c):
                return False
            node = node.children[c]
        return True


# Your Trie object will be instantiated and called as such:
obj = Trie()
obj.insert("hello")
obj.insert("hell")
print(obj.search("hello"))
print(obj.startsWith("he"))
print(obj.remove("hell"))
print(obj.startsWith("hello"))
# param_3 = obj.startsWith(prefix)
# @lc code=end

File: RustSolutions/src/208.implement-trie-prefix-tree.rs
/*
 * @lc app=leetcode id=208 lang=rust
 *
 * [208] Implement Trie (Prefix Tree)
 */

// @lc code=start
use std::collections::HashMap;

struct Trie {
    children: HashMap<char, Trie>,
    is_word: bool,
    refs: i32,
}

impl Trie {
    fn new() -> Self {
        Trie {
            children: HashMap::new(),
            is_word: false,
            refs: 0,
        }
    }

    fn insert(&mut self, word: String) {
        let mut node = self;
        for c in word.chars() {
            node = node.children.entry(c).or_insert(Trie::new());
        }
        node.is_word = true;
    }

    // fn remove(&mut self, word: &str) {
    //     let mut path: Vec<&mut Trie> = Vec::new();
    //     path.push(self);

    //     for c in word.chars() {
    //         if let Some(n) = path[path.len() - 1].children.get_mut(&c) {
    //             path.push(n);
    //         } else {
    //             return;
    //         }
    //     }

    //     let n = path.len();
    //     if !path[n - 1].is_word {
    //         return;
    //     }
    //     path[n - 1].is_word = false;
    //     if path[n - 1].children.len() > 0 {
    //         return;
    //     }

    //     let mut remroot = word.chars().last().unwrap();
    //     for (node, c) in path[0..n - 1].iter_mut().zip(word.chars().rev().skip(1)) {
    //         if node.children.len() > 1 || node.is_word {
    //             node.children.remove(&remroot);
    //             return;
    //         }
    //         remroot = c;
    //     }
    // }

    fn search(&self, word: String) -> bool {
        let mut node = self;
        for c in word.chars() {
            if let Some(n) = node.children.get(&c) {
                node = n;
            } else {
                return false;
            }
        }
        return node.is_word;
    }

    fn starts_with(&self, prefix: String) -> bool {
        let mut node = self;
        for c in prefix.chars() {
            if let Some(n) = node.children.get(&c) {
                node = n;
            } else {
                return false;
            }
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * let obj = Trie::new();
 * obj.insert(word);
 * let ret_2: bool = obj.search(word);
 * let ret_3: bool = obj.starts_with(prefix);
 */
// @lc code=end

